File: src/main/python/ghost.py
from direction import avoid, next_location, wrap, left, right, opposite
from direction import Direction
from random import randint
from game_element import GameElement
from colour import Colour


class Ghost(GameElement):
    icons = {
            "M": False,
            "W": True
            }
    icon_state = {
            False: {"icon": "M", "colour": None},
            True: {"icon": "W", "colour": Colour.BLUE}
            }

    colours = [Colour.RED,
               Colour.GREEN,
               Colour.PURPLE,
               Colour.CYAN]

    colour = 0

    def is_ghost(icon):
        if isinstance(icon, Ghost) or icon in Ghost.icons:
            return True
        return False

    def get_element(coords, icon):
        if (Ghost.is_ghost(icon)):
            return Ghost(coords, icon)

    def __init__(self, coordinates, icon=""):
        super(Ghost, self).__init__(coordinates, icon)
        self.start = coordinates
        self.passed_gate = False
        self.orig_colour = Ghost.colours[Ghost.colour]
        Ghost.colour = (Ghost.colour + 1) % len(Ghost.colours)

        if icon in Ghost.icons:
            self.icon = icon
        self.panic_level = 0 if Ghost.icons[icon] is False else 50
        self.direction = Direction.UP

    def add_to_game(self, game):
        self.game = game
        self.width = game.field.width()
        self.height = game.field.height()
        game.add_ghost(self)

    def panicked(self):
        return (self.panic_level > 0)

    def panic(self):
            self.panic_level = 50
            for icon, panic in Ghost.icons.items():
                if panic is True:
                    self.icon = icon

    def trigger_effect(self, pill):
        if pill.score() == 50:
            self.panic()

    def tick(self):
        self.manage_panic()
        self.choose_direction()
        if (self.game.is_pacman(self.coordinates)):
            if (self.panicked() is True):
                self.game.kill_ghost(self)
            else:
                self.game.kill_pacman()
        if (self.game.is_gate(self.coordinates)):
            self.passed_gate = True
        image = Ghost.icon_state[self.panic_level > 0]
        self.icon = image["icon"]
        this_colour = image["colour"]
        self.colour = self.orig_colour if this_colour is None else this_colour

    def manage_panic(self):
        if (self.panicked() is True):
            pacman_loc = self.coordinates
            pacman = self.game.get_pacman()
            if (pacman):
                pacman_loc = pacman.coordinates
            self.direction = avoid(self.coordinates, pacman_loc)
            self.panic_level -= 1

    def choose_direction(self):
        front = self.direction
        choices = [front, left(front), right(front)]
        options = self.find_options(choices)
        if len(options) > 0:
            self.random_choice(options)
        else:
            self.no_option()
        self.move()

    def find_options(self, choices):
        options = []
        for option in choices:
            if self.clear(self.next_step(option)):
                options.append(option)
        return options

    def random_choice(self, options):
        self.direction = options[randint(0, len(options)-1)]

    def no_option(self):
        if self.panicked() is False:
            self.direction = opposite(self.direction)

    def move(self):
        next_location = self.next_step(self.direction)
        if ((self.panic_level % 2) == 0) and self.clear(next_location):
            self.coordinates = (next_location)

    def next_step(self, direction):
        return wrap(next_location(self.coordinates, direction),
                    self.width,
                    self.height)

    def clear(self, coordinates):
        if not self.game.is_wall(coordinates):
            return True
        if self.game.is_gate(coordinates) and not self.passed_gate:
            return True
        return False

    def kill(self):
        self.coordinates = self.start
        self.panic_level = 0
        self.passed_gate = False

    def score(self):
        return 200

File: src/main/node/ghost.js
"use strict";

const d = require("./direction.js");
const nextLocation = d.nextLocation;
const turnLeft = d.turnLeft;
const turnRight = d.turnRight;
const opposite = d.opposite;
const avoid = d.avoid;
const RIGHT = d.Direction.RIGHT;
const LEFT = d.Direction.LEFT;
const UP = d.Direction.UP;
const DOWN = d.Direction.DOWN;
const Colour = require("./colour.js");
const GameElement = require("./game-elements.js");

const icons = [
    {icon:"M",panicked:false,colour:null},
    {icon:"W",panicked:true,colour:Colour.BLUE}
  ];

const ghostColours = [Colour.RED,
                      Colour.GREEN,
                      Colour.PURPLE,
                      Colour.CYAN];
var ghostColour = 0;

function setGhostColour() {
  return ghostColours[ghostColour++];
}

function findPanicLevel(icon) {
  return (icons.find( (ghost) => {
          return ghost.icon === icon;
          }).panicked) ? 50 : 0;
}

function findIcon(panic) {
    return (icons.find( (ghost) => {
            return panic === ghost.panicked;
            }).icon);
  }

module.exports =

class Ghost extends GameElement {
  constructor (options) {
    super(options);
    if (options) {
      this.start = options.coords;
      this.points = 200;
      this.direction = UP;
      this.game = null;
      this.gatePassed = false;
      this.uniqueColour = setGhostColour();
      this.panicLevel = findPanicLevel(options.icon);
    }
  } 
  
  static isGhost(token) {
    return (icons.filter((element) => element.icon === token).length>0);
  }
  
  getElement(coords, icon) {
    if (Ghost.isGhost(icon)) {
      return new Ghost({coords, icon});
    }
  }
  
  addToGame(game) {
    game.addGhost(this);
  }
  
  colour() {
    return this.getColour(this.isPanicked());
  }
  
  getColour(panicked) {
    let colour= icons.find((a) => { return a.panicked === panicked;}).colour;
    if (colour) {
      return colour;
    } else {
      return this.uniqueColour;
    }
  }
  
  tick() {
    this.managePanic();
    this.chooseDirection();
    this.checkCollisions();
    this.image = findIcon(this.isPanicked());
  }
  
  managePanic() {
    if (this.isPanicked()) {
      let pacman=this.game.getPacman();
      let pacLoc= (pacman) ? pacman.getLocation() : this.getLocation();
      this.direction = avoid(this.getLocation(),pacLoc);
      this.panicLevel--;      
    } 
  }
  
  chooseDirection() {
    let choices = [this.direction, turnLeft(this.direction),turnRight(this.direction)];
    let options = this.findOptions(choices);
    this.direction = (options.length > 0) ? this.randomChoice(options) : this.noOption();
    this.move();
  }
  
  findOptions(choices) {
    let options = [];
    choices.forEach( (direction) => {
      let next = nextLocation(this.getLocation(), direction);
      if (this.isClear(next)) {
        options.push(direction);
      } 
    });
    return options;
  }
  
  noOption() {
    return (!this.isPanicked()) ? opposite(this.direction) : this.direction;
  }
  
  randomChoice(options) {
    let i = Math.floor((Math.random() * options.length));
    return options[i];
  }
  
  move() {
    if ((this.panicLevel % 2) === 0) {
      let next = nextLocation(this.getLocation(),this.direction);
      if (this.isClear(next))  {
        this.setLocation(next);
      }
    }
  }
  
  checkCollisions() {
    if (this.isOnPacman()) {
      this.collision(this.game.playfield.getLocation(this.getLocation()));
    }

    if (this.game.isGate(this.getLocation())) {
      this.gatePassed = true;
    }
  }
  
  isClear(loc) {
    return (this.isNotAWall(loc) || this.isOpenGate(loc)) ? true : false;
  }
  
  isNotAWall(loc) {
    return (!(this.game.isWall(loc)));
  }
  
  isOpenGate(loc) {
    let it = this.game.playfield.getLocation(loc);
    return (this.game.isWall(loc)) ? (it.isGate() && !this.gatePassed) : false;
  }
  
  restart() {
    this.setLocation(this.start);
    this.panicLevel=0;
    this.gatePassed = false;
  }
  
  collision(pacman) {
    if (this.isPanicked()) {
      this.game.addScore(this.score());
      this.restart();
    } else {
      pacman.kill();
    }
  }
  
  isOnPacman() {
    if (this.game.getPacman()) {
      let pacLoc = this.game.getPacman().getLocation();
      let ourLoc = this.getLocation();
      return (pacLoc.x === ourLoc.x && pacLoc.y === ourLoc.y);
    } else {
      return false;
    }
  }
  
  score() {
    return this.points;
  }
  
  panic() {
    this.panicLevel = 50;
    this.image = findIcon(this.isPanicked());
  }
  
  isPanicked() {
    return (this.panicLevel > 0);
  }
  
  setGame(game) {
    this.game = game;
  }
}